### PR TITLE
TST, MAINT: Lots of new tests for fromnumeric.py

### PR DESCRIPTION
**Pull request by gfyoung**
Follow-up from #7000 in which I found there seemed to be a gap in testing for the `numpy/numpy/core/fromnumeric.py` file.  This PR improves the coverage on the methods defined in that file.

### CODE REVIEW

**Comment by gfyoung** [comment]
PR is no longer WIP.  Travis + Appveyor are both passing.  Should be good to merge now.

**Comment by charris** [comment]
Please don't rebase.

**Comment by gfyoung** [comment]
PR is no longer WIP.  Travis + Appveyor are both passing.  Should be good to merge now.

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -4979,29 +5037,69 @@ def test_2d(self):
...
         self.x = 2*np.ones((3,), dtype=int)
         self.y = 3*np.ones((3,), dtype=int)
         self.x2 = 2*np.ones((2, 3), dtype=int)
         self.y2 = 3*np.ones((2, 3), dtype=int)
-        self.ind = [0, 0, 1]
+        self.ind = np.array([0, 0, 1])
 
     def test_basic(self):
-        A = np.choose(self.ind, (self.x, self.y))
+        A = self.ind.choose((self.x, self.y))
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: Should keep the first version in order to test the funtion, not the method.

Comment (charris) [inline numpy/core/tests/test_multiarray.py]: Same below.

File: numpy/core/tests/test_multiarray.py [hunk 2]
```diff
@@ -4979,29 +5037,69 @@ def test_2d(self):
...
         assert_equal(A, [[2, 2, 3], [2, 2, 3]])
 
 
+class TestRepeat(TestCase):
+    def setUp(self):
+        self.m = np.array([1, 2, 3, 4, 5, 6])
+        self.m_rect = self.m.reshape((2, 3))
+
+    def test_basic(self):
+        A = self.m.repeat([1, 3, 2, 1, 1, 2])
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: See previous comment.

File: numpy/lib/tests/test_function_base.py [hunk 3]
```diff
@@ -409,14 +409,13 @@ def test_basic(self):
             a = np.array(ba, ctype)
             a2 = np.array(ba2, ctype)
             if ctype in ['1', 'b']:
-                self.assertRaises(ArithmeticError, np.prod, a)
-                self.assertRaises(ArithmeticError, np.prod, a2, 1)
-                self.assertRaises(ArithmeticError, np.prod, a)
+                self.assertRaises(ArithmeticError, a.prod)
```
Comment (charris) [inline numpy/lib/tests/test_function_base.py]: See above about methods vs functions.

**Comment by charris** [comment]
Looks generally good, but the common mistake is to substitute methods for functions. The two aren't the same despite the apparent redundancy. The methods should actually be tested in the `numpy/core/tests/test_multiarray.py` huge `TestMethods` class. Note that Numeric was the first version of what became Numpy and didn't have many methods.

File: numpy/core/tests/test_multiarray.py (hunk 1 above)
Comment (gfyoung) [inline numpy/core/tests/test_multiarray.py]: Done.

File: numpy/core/tests/test_multiarray.py (hunk 2 above)
Comment (gfyoung) [inline numpy/core/tests/test_multiarray.py]: Done.

File: numpy/lib/tests/test_function_base.py (hunk 3 above)
Comment (gfyoung) [inline numpy/lib/tests/test_function_base.py]: Done.

**Comment by gfyoung** [comment]
@charris : Changes made and Travis / Appveyor are happy.

**Comment by gfyoung** [comment]
@charris : Changes made and Travis / Appveyor are happy.

File: numpy/core/tests/test_fromnumeric.py [hunk 4]
```diff
@@ -1,17 +0,0 @@
-from __future__ import division, absolute_import, print_function
```
Comment (charris) [inline numpy/core/tests/test_fromnumeric.py]: Why remove this file? `fromnumeric.py` is a package containing a number of functions. `put` is also a function implemented in `fromnumeric.py`.

Comment (gfyoung) [inline numpy/core/tests/test_fromnumeric.py]: I added that file myself from #7000 that served as a stopgap measure to test that ```put``` would not accept not ```ndarray``` args.  Also, a lot of the tests of ```fromnumeric.py``` functions are located elsewhere in the tests.

**Comment by charris** [comment]
I think you need to check carefully where functions/methods etc are defined. The easy way to check functions in ipython is to use the `?` help option on the function name. Or you can look at the `__all__` list in the modules to see what functions are exported. You can check ndarray methods with `ndarray.<themethod>?`.

**Comment by gfyoung** [comment]
If you look at the functions defined in ```fromnumeric.py```, almost all of them are tested elsewhere in the code-base, even though they are in the ```__all__``` array.  For example ```np.prod``` is defined <a href="https://github.com/numpy/numpy/blob/master/numpy/core/fromnumeric.py#L26">here</a>, but the test is <a href="https://github.com/numpy/numpy/blob/54224f4e88036a8cd02c1306f2c1a5f655e41e3a/numpy/lib/tests/test_function_base.py#L402">here</a>

My original intention was to write an entire ```test_fromnumeric.py``` but after searching the code base for many of these functions, I realized that I would just be duplicating already existing tests.

File: numpy/core/tests/test_fromnumeric.py (hunk 4 above)
Comment (charris) [inline numpy/core/tests/test_fromnumeric.py]: Hmm, at some point we need to reorganize `numpy/core`, but I suppose that is for another day.

**Comment by charris** [comment]
Thanks @gfyoung .

**Pull request merged**